src/main/java/com/google/api/codegen/configgen/CollectionPattern.java

```diff
@@ -34,45 +38,136 @@ public static CollectionPattern create(FieldSegment fieldSegment) {
     if (!isValidCollectionPattern(fieldSegment)) {
       throw new IllegalArgumentException("Field segment is not a valid collection pattern.");
     }
-    return new CollectionPattern(fieldSegment);
+
+    Stream<PathSegment> wildcardSegments =
+        fieldSegment.getSubPath().stream().filter(segment -> segment instanceof WildcardSegment);
+    String simpleName =
+        Streams.findLast(wildcardSegments)
+            .map(segment -> fieldSegment.getSubPath().indexOf(segment))
+            .filter(index -> index >= 0)
+            .map(index -> buildEntityName(fieldSegment.getSubPath(), index))
+            .orElseThrow(() -> new IllegalStateException("Field segment contained no wildcards."));
+
+    List<PathSegment> templatizedSubpath =
+        Streams.mapWithIndex(
+                fieldSegment.getSubPath().stream(),
+                (segment, index) ->
+                    buildTemplatizedSegment(segment, (int) index, fieldSegment.getSubPath()))
+            .collect(toImmutableList());
+
+    return new CollectionPattern(fieldSegment, templatizedSubpath, simpleName);
   }
 
-  private final FieldSegment fieldSegment;
-  private final List<PathSegment> templatizedSubpath;
-  private final String simpleName;
+  /**
+   * Builds a templatized path segment from the specified segment, replacing a wildcard with a
+   * literal.
+   *
+   * @param segment the segment to replace
+   * @param index the index of the segment in segments
+   * @param pathSegments the full path containing segment at index
+   * @return a templatized path segment with any wildcard segment replaced with a literal segment
+   *     containing the entity name
+   */
+  private static PathSegment buildTemplatizedSegment(
+      PathSegment segment, int index, List<PathSegment> pathSegments) {
+    if (!(segment instanceof WildcardSegment)) {
+      return segment;
+    }
 
-  private CollectionPattern(FieldSegment fieldSegment) {
-    this.fieldSegment = fieldSegment;
+    String prefix = buildSimpleEntityPrefix(pathSegments, index);
+    String entityName = buildEntityName(pathSegments, index, prefix);
+    if (((WildcardSegment) segment).isUnbounded()) {
```

Comment:
The prior function call to `buildEntityName` has only one purpose, to check for isUnbounded and append `_path`. WDYT about removing the call to `buildEntityName` from here and appending `_path` before constructing the LiteralSegment? This results in duplicating the logic to append `_path`, but unduplicates the checking for isUnbounded, and means that the function `buildEntityName` function could likely be completely removed. I'm not that committed to this change though, if you prefer as-is then that's fine.

Reply:
Done